File: SDCLinuxExpansionBoardDriver_V2.1.4.0/GPIO/python/gpio_test.py
#! /usr/bin/python3
import gpiod
import time
import os
import keyboard
import re
import multiprocessing
import click

from gpiod.line import Direction, Value, Edge
from gpiod.edge_event import EdgeEvent
from gpiod import LineRequest

GPIO_CLASS_PATH = "/dev/"

gpio_chip_device_name = None
gpio_chip = None

w_proc = None
is_waiting = False

def GpioTestList():
 
    os.system('clear')
    ListGpioChips()
    print('>------------------------------------<')
    print('1 : GPIO Chip Open')
    print('2 : GPIO Chip Close')
    print('3 : Get Chip Line List')
    print('4 : Set Output Value')
    print('5 : Get Input Value')
    print('6 : Wait Input Event')
    print()
    print('? : Command list')
    print('Q : Quit')
    print('>------------------------------------<')

def ListGpioChips():
    
    print("=======================================================")
    print("Available GPIO chips in the system:")
    print("Name\t\tLabel\t\t\t\tLines")

    # Loop through all files in the /dev/ directory and look for gpiochip*
    for entry in os.listdir(GPIO_CLASS_PATH):
        if entry.startswith("gpiochip"):
            chip_path = os.path.join(GPIO_CLASS_PATH, entry)
            
            try:
                chip = gpiod.Chip(chip_path)
                label = chip.get_info().label or "[No Label]"
                num_lines = chip.get_info().num_lines

                print(f"{entry}\t{label:<30}\t{num_lines}")
            except Exception as e:
                print(f"{entry}\t\t[Failed to open] ({e})")

    print("=======================================================")

def GpioChipOpen():

    global gpio_chip_device_name
    global gpio_chip

    print('GPIO device number(e.g if /dev/gpiochip1, input 1)?')
    ch_num = input()

    if re.search('^[0-9]+$', ch_num):

        gpio_chip_device_name = f"/dev/gpiochip{ch_num}"

        try:
            gpio_chip = gpiod.Chip(gpio_chip_device_name)
            print(f"Open {gpio_chip_device_name} Successfully")

        except OSError:
            print(f"Error opening GPIO chip: {gpio_chip_device_name}")          
    else:
        print('Error: Wrong GPIO device number input')

def GpioChipClose():

    global gpio_chip_device_name
    global gpio_chip

    if gpio_chip:

        gpio_chip.close()
        print(f"{gpio_chip_device_name} closed")

def GetChipLines():

    global gpio_chip_device_name
    global gpio_chip

    num_lines = gpio_chip.get_info().num_lines

    print(f"==========={gpio_chip_device_name} lines:{num_lines}===========")
    for i in range(num_lines):
        try:
            info = gpio_chip.get_line_info(i)     
            direction = info.direction.name if info.direction else "unknown"
            print(f"line{i}\t direction: {direction}")

        except Exception as e:
            print(f"Error to get: line{i} ({e})")
    print("=============================================")
    

def GpioChipSetOutput():
    
    try:
        
        global gpio_chip_device_name
        global gpio_chip

        print('GPIO chip line number  ?')
        line_num = input()

        if re.search('^[0-9]+$', line_num):

            print('Output value(0 or 1)?')
            value = input()
        
            if re.fullmatch(r'[01]', value):
                
                with gpio_chip.request_lines(config={line_num: gpiod.LineSettings(direction=Direction.OUTPUT, output_value=Value.INACTIVE)}, consumer='gpio_output') as request: 
                   if value=='1': 
                       request.set_value(line_num, Value.ACTIVE)
                       print(f'{gpio_chip_device_name} line{line_num} set value:1')

                   else:
                       request.set_value(line_num, Value.INACTIVE)
                       print(f'{gpio_chip_device_name} line{line_num} set value:0')
            else:
                print('Error: Wrong output value input')
        else:
            print('Error: Wrong line number input')
            
    except Exception as e:
        print(f"GPIO error: {e}")

def GpioChipGetInput():

    try:
        
        global gpio_chip_device_name
        global gpio_chip

        print('GPIO chip line number  ?')
        line_num = input()

        if re.search('^[0-9]+$', line_num):

            with gpio_chip.request_lines(config={line_num: gpiod.LineSettings(direction=Direction.INPUT)}, consumer='gpio_input') as request:
            
                value = request.get_value(line_num)
                           
                if value == Value.INACTIVE:    
                
                    print(f'{gpio_chip_device_name} line{line_num} get value:0')
                
                else:

                    print(f'{gpio_chip_device_name} line{line_num} get value:1')
        else:
                print('Error: Wrong line number input')
    except Exception as e:
        print(f"GPIO error: {e}")

def WaitInputEvent():

    try:
        
        global gpio_chip_device_name
        global gpio_chip

        print('GPIO chip line number  ?')
        line_num = input()

        if re.search('^[0-9]+$', line_num):
            
            with gpio_chip.request_lines(config={line_num: gpiod.LineSettings(direction=Direction.INPUT, edge_detection=Edge.BOTH)}, consumer='gpio_event') as request:
                
                global w_proc
                w_proc = multiprocessing.Process(target=EventWait, daemon=True, args=(request, line_num,)) # Multiprocessing replace threading, cus easy to terminate the process

                w_proc.start()
                print(f'Waiting line{line_num} event...press "esc" key to exit.')           
                keyboard.wait('esc')           
                w_proc.terminate()
        else:
                print('Error: Wrong line number input')
    
    except Exception as e:
        print(f"GPIO error: {e}")

def EventWait(request: LineRequest, line_num):
    
    while True:

        try:
        
            if request.wait_edge_events(timeout=0.5):

                event_list = request.read_edge_events(max_events=5)
                    
                for e in event_list:

                    if e.event_type == EdgeEvent.Type.RISING_EDGE:
                        print(f"line{line_num} rising edge triggered. Offset:{e.line_offset}, Timestamp:{e.timestamp_ns/1000000000}")
                
                    elif e.event_type == EdgeEvent.Type.FALLING_EDGE:
                        print(f"line{line_num} falling edge triggered. Offset:{e.line_offset}, Timestamp{e.timestamp_ns/1000000000}")
     

        except Exception as e:
            print(f"GPIO error: {e}")


def main():
    
    GpioTestList()

    while True: 

        print()
        print('CMD?')

        key = click.getchar()

        if(key == '1'):

            GpioChipOpen()
        
        elif(key == '2'):

            GpioChipClose()
        
        elif(key == '3'):

            GetChipLines()
        
        elif(key == '4'):

            GpioChipSetOutput()
        
        elif(key == '5'):

            GpioChipGetInput()
        
        elif(key == '6'):

            WaitInputEvent()

        elif(key == '?'):
                     
            GpioTestList()
            
        elif(key == 'Q'or  key == 'q'):
      
            quit()

        time.sleep(0.2)

if __name__ == '__main__':
    main()

File: SDCLinuxExpansionBoardDriver_V2.1.4.0/GPIO/c/gpio_test/gpio_test.c
/**
 * @file gpio_test.c
 * @brief cmd list for test GPIO function
 * 
 * SPDX-License-Identifier: GPL-2.0-only
 *
 * Copyright (C) 2025, SUNIX Co., Ltd.
 *
 * Authors      : Max Chang <max.chang@sunix.com>
 * 
 * 
**/

#include "precomp.h"

#define ESC_KEY 27
#define GPIO_CLASS_PATH "/dev/"

volatile int running = 1; 

char gpio_chip[20];
struct gpiod_chip *chip;
struct gpiod_line *line;
struct gpiod_line_event event;

struct timespec timeout;

struct termios oldt;

void set_terminal_mode() {
	struct termios newt;
    tcgetattr(STDIN_FILENO, &oldt);  
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);  
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
}

void reset_terminal_mode() {
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
}

void *keyboard_listener(void *arg) {
    char key;
    while (running) {
        if (read(STDIN_FILENO, &key, 1) > 0) {
            if (key == ESC_KEY) {
                printf("\nStop wating GPIO event...\n");
				running = 0;
                break;
            }
        }
        usleep(50000); 
    }
    return NULL;
}

static void _print_gpio_cmd(void)
{
	printf(">>------------------------------------<<\n");
	printf("     A : GPIO Chip Open\n");
	printf("     B : GPIO Chip Close\n");
	printf("     C : Get Chip Line List\n");
	printf("     D : Set Output Value\n");
	printf("     E : Get Input Value\n");
    printf("     F : Wait Input Event\n");
	printf("\n");
	printf("     Q : quit\n");
	printf(">>------------------------------------<<\n\n");
}

void list_gpio_chips()
 {
    struct dirent *entry;
    DIR *dir = opendir(GPIO_CLASS_PATH);

    if (!dir) {
        perror("Failed to open GPIO directory");
        return;
    }

	printf("=======================================================\r\n");
    printf("Available GPIO chips in the system:\r\n");
    printf("Name\t\tLabel\t\t\t\tLines\r\n");

    while ((entry = readdir(dir)) != NULL) {
        // Check if the entry is a gpiochip (e.g., gpiochip0, gpiochip1, etc.)
        if (strncmp(entry->d_name, "gpiochip", 8) == 0) {
            char chip_path[256];
            snprintf(chip_path, sizeof(chip_path), "%s%.250s", GPIO_CLASS_PATH, entry->d_name);

            struct gpiod_chip *chip = gpiod_chip_open(chip_path);
            if (!chip) {
                printf("%s\t\t[Failed to open]\n", entry->d_name);
                continue;
            }

            const char *label = gpiod_chip_label(chip);
            int num_lines = gpiod_chip_num_lines(chip);
            
            printf("%s\t%-30s\t%d\n", entry->d_name, label ? label : "[No Label]", num_lines);

            gpiod_chip_close(chip);
        }
    }
	printf("=======================================================\r\n");
    closedir(dir);
}

void gpio_chip_open()
{
	int n = input_gpio_device_number();
		
	sprintf(gpio_chip, "/dev/gpiochip%d", n);

	chip = gpiod_chip_open(gpio_chip);
	
	if (!chip) {
        printf("Error opening GPIO chip: %s\r\n", gpio_chip);
    }
	else
	{
		printf("Open %s Successfully\r\n", gpio_chip);
	}
}

void gpio_chip_close()
{
	if(chip)
	{
		gpiod_chip_close(chip);
		printf("%s closed\r\n",gpio_chip);
	}
}

void get_chip_lines()
{
	int num_lines = gpiod_chip_num_lines(chip);

	printf("========== %s:%d lines ==========\r\n", gpio_chip, num_lines);
	for (int i = 0; i < num_lines; i++) 
	{
		struct gpiod_line *line = gpiod_chip_get_line(chip, i);

		int direction = gpiod_line_direction(line);

		const char *dir = (direction == GPIOD_LINE_DIRECTION_INPUT) ? "Input" : (direction == GPIOD_LINE_DIRECTION_OUTPUT) ? "Output" : "Unkown";

		if (line) {			
			printf("line%d\t direction:%s\n", i, dir);
		} else {
			printf("Error to get: line%d\n", i);
		}

		gpiod_line_release(line);
	}
	printf("=============================================");
}

void gpio_chip_set_output()
{
	if(chip)
	{
		int line_num = input_line_number();
		line = gpiod_chip_get_line(chip, line_num);
		
		if(!line)
		{
			printf("Error get GPIO chip line: %d\r\n", line_num);
			gpiod_chip_close(chip);
		}

		if(gpiod_line_request_output(line, "gpio_output", 0) < 0)
		{
			printf("Error: gpiod_line_request_output\r\n");
			gpiod_chip_close(chip);
		}

		int value = input_gpio_output_level_value();

		int result = gpiod_line_set_value(line, value);

		if(result != 0)
		{
			printf("Set output value fail, line:%d, GPIO chip:%s\r\n", line_num, gpio_chip);
		}
		else
		{
			printf("Set output value successfully, line:%d, GPIO chip:%s\r\n", line_num, gpio_chip);
		}

		gpiod_line_release(line);
	}
	else
	{
		printf("Error: gpiod_chip is null\r\n");
	}
}

void gpio_chip_get_input()
{
	if(chip)
	{
		int line_num = input_line_number();
		line = gpiod_chip_get_line(chip, line_num);

		if(!line)
		{
			printf("Error get GPIO chip line: %d\r\n", line_num);
			gpiod_chip_close(chip);
		}

		if (gpiod_line_request_input(line, "gpio_input") < 0) 
		{
			printf("Error: gpiod_line_request_input\r\n");
			gpiod_chip_close(chip);
		}

		int value = gpiod_line_get_value(line);

		if(value < 0)
		{
			printf("Get input value fail, line:%d, GPIO chip:%s\r\n", line_num, gpio_chip);
		}
		else
		{
			printf("line%d input value:%d\r\n", line_num,value);
		}

		gpiod_line_release(line);
	}
	else
	{
		printf("Error: gpiod_chip is null\r\n");
	}
}

void gpio_chip_wait_input_event()
{	
	timeout.tv_sec = 0.5;
	timeout.tv_nsec = 0;

	if(chip)
	{
		int result;

		int line_num = input_line_number();
		line = gpiod_chip_get_line(chip, line_num);

		if(!line)
		{
			printf("Error get GPIO chip line: %d\r\n", line_num);
			gpiod_chip_close(chip);
		}
		
		if (gpiod_line_request_both_edges_events(line, "gpio_event") < 0)
		{
			printf("Error: gpiod_line_request_both_edge_events\r\n");
			gpiod_chip_close(chip);
		}
		
		set_terminal_mode();
		pthread_t thread;
        pthread_create(&thread, NULL, keyboard_listener, NULL);

		printf("Waiting line%d event...press 'Esc' to exit\r\n", line_num);

		running = 1;

		while (running)
		{

			result = gpiod_line_event_wait(line, &timeout);
			
			if(result > 0)
			{
				gpiod_line_event_read(line, &event);
				
				switch (event.event_type)
				{
					case GPIOD_LINE_EVENT_RISING_EDGE:
						
						printf("line%d rising edge triggered, timestamp:%" PRIdMAX ".%09" PRIdMAX "\r\n",line_num, (intmax_t)event.ts.tv_sec, (intmax_t)event.ts.tv_nsec);
						break;
					
					case GPIOD_LINE_EVENT_FALLING_EDGE:
	
						printf("line%d falling edge triggered, timestamp:%" PRIdMAX ".%09" PRIdMAX "\r\n",line_num, (intmax_t)event.ts.tv_sec, (intmax_t)event.ts.tv_nsec);
						break;
				}

			}
		}
		pthread_join(thread, NULL);
		gpiod_line_release(line);
		reset_terminal_mode();
	}
	else
	{
		printf("Error: gpiod_chip is null\r\n");
	}
}

void gpio_test(void)
{
    int cmd;

    (void)!system("clear");

	list_gpio_chips();

	_print_gpio_cmd();

    do{
		printf("\n>> CMD ?\n");
		cmd = sgetche();
		printf("\n");

		switch (cmd) {
		case 'a':
		case 'A':
			gpio_chip_open();
			break;
		case 'b':
		case 'B':
			gpio_chip_close();
			break;
		case 'c':
		case 'C':
			get_chip_lines();
			break;
		case 'd':
		case 'D':
			gpio_chip_set_output();
			break;
		case 'e':
		case 'E':
			gpio_chip_get_input();
			break;
		case 'f':
		case 'F':
			gpio_chip_wait_input_event();
			break;
		case '?':
			_print_gpio_cmd();
			break;
		}

    }while (cmd != 'q' && cmd != 'Q');
}

